### PR TITLE
OpenSLR Javanese ASR

### PR DESCRIPTION
**Pull request by peter-yh-wu**
ASR for [OpenSLR Large Javanese ASR training data set](http://www.openslr.org/35/)

### CODE REVIEW

File: egs2/java/asr1/local/data.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+. utils/parse_options.sh
+
+log "data preparation started"
+
+mkdir downloads
```
Comment (ftshijt) [inline egs2/java/asr1/local/data.sh:34]: Better to register the download directory in `db.sh`. It also helps if users want to download data in another dir.

File: egs2/java/asr1/local/data.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    python3 local/java_data_prep.py
+    utils/spk2utt_to_utt2spk.pl data/train/spk2utt > data/train/utt2spk
+    utils/spk2utt_to_utt2spk.pl data/dev/spk2utt > data/dev/utt2spk
+    utils/spk2utt_to_utt2spk.pl data/test/spk2utt > data/test/utt2spk
+fi
+
+if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
+    log "stage 2: Create Non-linguistic Symbols for Language ID"
+    mkdir -p data/local
+    touch ${nlsyms_txt}
```
Comment (ftshijt) [inline egs2/java/asr1/local/data.sh:59]: I find that you want to use the score_lang_id, but I do not find where you add language labels to the text. If you already add them somewhere, please ignore this comment.

File: egs2/java/asr1/local/java_data_prep.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        if os.path.exists(path):
+            utt2text[fid] = text
+            if spk in spk2utt:
+                spk2utt[spk].append(fid)
+            else:
+                spk2utt[spk] = [fid]
+
+    spks = sorted(list(spk2utt.keys()))
+    random.Random(0).shuffle(spks)
+    num_train = int(len(spks) * 0.8)
```
Comment (ftshijt) [inline egs2/java/asr1/local/java_data_prep.py:33]: I found this in the original paper for this dataset, maybe it would be better to follow their split? We may have some utterances from training samples as a development set.
```

We followed an identical procedure for all languages: The
full dataset was divided into two subsets, one for training, and
the other for testing. The test data was produced by grouping
the data by SpeakerID and holding the first 2000 lines out;
the remainder was used for training.
```

File: egs2/java/asr1/local/data.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+fi
+
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+    log "sub-stage 1: Preparing Data for openslr"
+
+    python3 local/java_data_prep.py
+    utils/spk2utt_to_utt2spk.pl data/train/spk2utt > data/train/utt2spk
+    utils/spk2utt_to_utt2spk.pl data/dev/spk2utt > data/dev/utt2spk
+    utils/spk2utt_to_utt2spk.pl data/test/spk2utt > data/test/utt2spk
+fi
```
Comment (ftshijt) [inline egs2/java/asr1/local/data.sh:54]: I would suggest to add `utils/fix_data_dir.sh` here, it's not mandatory but would be better.

File: egs2/java/asr1/local/data.sh (hunk 2 above)
Comment (ftshijt) [reply egs2/java/asr1/local/data.sh:59]: The final text of multi-lingual sets with language id should be like:
```

utt_id1 [lang_id1] text
utt_id2 [lang_id2] text
...
```

**Review by ftshijt — COMMENTED** [review]
Thank you very much. The recipe generally looks good. Just some comments as following:

File: egs2/TEMPLATE/asr1/db.sh [hunk 5]
```diff
@@ -71,6 +71,7 @@ YESNO=downloads
 HOW2_TEXT=downloads/how2-300h-v1
 HOW2_FEATS=downloads/fbank_pitch_181516
 ZEROTH_KOREAN=downloads
+JAVA=downloads
```
Comment (ftshijt) [inline egs2/TEMPLATE/asr1/db.sh:74]: please also add one to clsp.jhu.edu (you can leave that empty)

File: egs2/java/asr1/conf/tuning/decode_transformer.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+batch_size: 1
```
Comment (ftshijt) [inline egs2/java/asr1/conf/tuning/decode_transformer.yaml:1]: why using 1 as batchsize? Wouldn't it be a bit slow, just curious.

File: egs2/java/asr1/local/data.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+log "data preparation started"
+
+mkdir -p ${JAVA}
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    log "sub-stage 0: Download Data to downloads"
+
+    wget https://www.openslr.org/resources/35/asr_javanese_0.zip
+    mv asr_javanese_0.zip downloads
```
Comment (ftshijt) [inline egs2/java/asr1/local/data.sh:39]: I think you need to change `downloads` to ${JAVA}

File: tools/activate_python.sh [hunk 8]
```diff
@@ -1,2 +1,6 @@
-#!/usr/bin/env bash
```
Comment (ftshijt) [inline tools/activate_python.sh:1]: Please keep the original activate_python.sh during PR

**Review by ftshijt — COMMENTED** [review]
Many thanks! just some minor comments as follows:

File: egs2/java/asr1/local/score.sh [hunk 9]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (ftshijt) [inline egs2/java/asr1/local/score.sh:1]: Seems the same as open_li52. If it is true, recommend using a softlink here

**Review by ftshijt — COMMENTED** [review]
Many thanks. It looks in a good shape now. If CI checks are successful, we can merge that

File: tools/activate_python.sh [hunk 10]
```diff
@@ -1,2 +1,2 @@
 #!/usr/bin/env bash
```
Comment (sw005320) [inline tools/activate_python.sh:1]: I think it should not be included in this PR.
Please remove it.

File: egs2/java/asr1/local/score.sh (hunk 9 above)
Comment (sw005320) [inline egs2/java/asr1/local/score.sh:1]: Do we need this?
I think we don't need this because this is a monolingual recipe.

**Review by sw005320 — COMMENTED** [review]
Yes, it looks good to me.
I have a couple of minor comments
- the name of this recipe should be refined `java` is not a good name. <some name of this corups>_jv or openslr35? Please find the correct name.
- please add a brief corpus description in https://github.com/espnet/espnet/blob/master/egs2/README.md
- can you also add a pre-trained model? @ftshijt, can you help @peter-yh-wu for this?

**Comment by ftshijt** [comment]
seems have conflict, can you fix that?

**Comment by ftshijt** [comment]
> seems have conflict, can you fix that?

I fixed that on my own. Nice job, Peter @peter-yh-wu.

**Comment by peter-yh-wu** [comment]
> > seems have conflict, can you fix that?
> 
> I fixed that on my own. Nice job, Peter @peter-yh-wu.

Thanks! :)
